### PR TITLE
add backend/gslice.d:  struct slice optimization

### PR DESCRIPTION
**Pull request by WalterBright**
This enables 'slicing' a two register wide aggregate into two register-sized variables, enabling much better enregistering. This is just a start for this. I'm doing this in smaller steps to make it easier to track down causes of any regressions.

### CODE REVIEW

**Comment by WalterBright** [comment]
Currently blocked by https://github.com/dlang/dmd/pull/6174

**Comment by WalterBright** [comment]
Given the code:
```
void foo(int[] a, int[] b, int[] c) {
    foreach (i; 0 .. a.length)
        a[i] = b[i] + c[i];
}
```
the inner loop currently compiles to:
```
LA:             mov     EAX,018h[ESP]
                mov     EDX,010h[ESP]
                mov     ECX,[EBX*4][EAX]
                add     ECX,[EBX*4][EDX]
                mov     ESI,020h[ESP]
                mov     [EBX*4][ESI],ECX
                inc     EBX
                cmp     EBX,01Ch[ESP]
                jb      LA
```
with this PR:
```
L1A:            mov     ECX,[EBX*4][EDI]
                add     ECX,[EBX*4][ESI]
                mov     0[EBX*4][EBP],ECX
                inc     EBX
                cmp     EBX,EDX
                jb      L1A
```

File: src/backend/gslice.c
Comment (andralex) [inline src/backend/gslice.c]: The complement is actually easier to write and read:

```d
if (tysize(e->Ety) != REGSIZE || e->Eoffset != 0 && e->Eoffset != REGSIZE)
{
    sia[si].canSlice = false;
}
```

File: src/backend/gslice.c
Comment (andralex) [inline src/backend/gslice.c]: switch with two branches is odd

File: src/backend/gslice.c
Comment (andralex) [inline src/backend/gslice.c]: so... why this scope here? looks suspect - remove or explain in a comment (I assume it's because of the goto and the variables)

File: src/backend/gslice.c
Comment (andralex) [inline src/backend/gslice.c]: The complement is actually easier to write and read:

```d
if (tysize(e->Ety) != REGSIZE || e->Eoffset != 0 && e->Eoffset != REGSIZE)
{
    sia[si].canSlice = false;
}
```

File: src/backend/gslice.c
Comment (andralex) [inline src/backend/gslice.c]:
```d
if (!sia[si].canSlice)
{
    continue;
}
... bunch of code ...
```

**Review by andralex — APPROVED** [review]
ok modulo some nits

File: src/backend/gslice.c
Comment (WalterBright) [inline src/backend/gslice.c]: ok

File: src/backend/gslice.c
Comment (WalterBright) [inline src/backend/gslice.c]: I expect to add more in the future as I make this optimization more capable.

File: src/backend/gslice.c
Comment (WalterBright) [inline src/backend/gslice.c]: ok

File: src/backend/gslice.c
Comment (WalterBright) [inline src/backend/gslice.c]: I like to avoid things like "not can slice", the negations are cognitively harder.

**Comment by WalterBright** [comment]
Auto-merge toggled on

File: src/backend/gslice.c
Comment (MartinNowak) [inline src/backend/gslice.c]: This body is complex, very low-level, does sth. unexpected (seems like it's replacing the symbol in the global symtab), and comes without explanation/comment. Please add at least a short sentence for the reasoning (what and why), instead of having readers spent 5min. to unpuzzle the low level code, and 10min. on understanding.

**Comment by andralex** [comment]
A note to everyone: the new review flow allows one to pull their own request if it was approved. This gives people the opportunity to mind the review nits before merging.

File: src/backend/gslice.c
Comment (MartinNowak) [inline src/backend/gslice.c]: FWIW si0 is always the same as the index in sia, so it seems redundant.

File: src/backend/gslice.c
Comment (MartinNowak) [inline src/backend/gslice.c]: A brief overview of what and this module does and how it does it would be really friendly.

File: src/backend/gslice.c
Comment (MartinNowak) [inline src/backend/gslice.c]: No idea what this Eoffset = 0 is trying to achieve.

File: src/backend/gslice.c
Comment (MartinNowak) [inline src/backend/gslice.c]: Why + 1?

**Comment by WalterBright** [comment]
Auto-merge toggled off

File: src/backend/gslice.c
Comment (WalterBright) [inline src/backend/gslice.c]: It's different, off by `n`.

File: src/backend/gslice.c
Comment (WalterBright) [inline src/backend/gslice.c]: I made it a struct so I can add more things to it later. It's very unlikely to get large, as it's the number of local symbols in a function, so size isn't a problem.

File: src/backend/gslice.c
Comment (WalterBright) [inline src/backend/gslice.c]: 3 is because it is used for two arrays, sia[] and sia2[]. sia2[] can grow to twice the size of sia[], as symbols can get split into two.

File: src/backend/gslice.c
Comment (WalterBright) [inline src/backend/gslice.c]: added comment

File: src/backend/gslice.c
Comment (WalterBright) [inline src/backend/gslice.c]: The second slice is the next symbol.

**Comment by WalterBright** [comment]
Auto-merge toggled on

File: src/backend/gslice.c
Comment (MartinNowak) [inline src/backend/gslice.c]: Well this is the only place where si0 gets set and it reads `sia2[si + n].si0 = si + n;`.

**Comment by MartinNowak** [comment]
Now that I read that you use the term slice not only for enregistering slice but for structs in general, I'd highly recommend to use a different term.
With the wide use of the term slice in D, reading this code is just confusing.

**Comment by WalterBright** [comment]
Yes, it's SROA: http://digital.cs.usu.edu/~allan/AdvComp/Notes/earlyd/

**Comment by MartinNowak** [comment]
According to Digger this caused [Issue 17215 – ICE(cgcod.c:findreg) with SIMD and -O -inline](https://issues.dlang.org/show_bug.cgi?id=17215).

**Pull request merged**
